File: implementations/react/src/patterns/behavioral/state/progress-state.ts
import { ClosedState } from "./closed-state";
import type { TicketContext } from "./context";
import { TicketState } from "./state";

export class InProgressState extends TicketState {
  private ctx: TicketContext;
  constructor(ctx: TicketContext) {
    super();
    this.ctx = ctx;
  }
  handleRequest() {
    const task = this.ctx.checkNextTask();
    if (task) {
      if (this.ctx.allDone()) {
        this.ctx.setState(new ClosedState(this.ctx));
      }
    } else {
      this.ctx.setState(new ClosedState(this.ctx));
    }
  }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/state/progress_state.dart
import 'package:design_patterns_flutter/patterns/behavioral/state/closed_state.dart';
import 'package:design_patterns_flutter/patterns/behavioral/state/context.dart';
import 'package:design_patterns_flutter/patterns/behavioral/state/state.dart';

class InProgressState extends TicketState {
  final TicketContext ctx;
  InProgressState(this.ctx);

  @override
  void handleRequest() {
    final task = ctx.checkNextTask();
    if (task != null) {
      if (ctx.allDone()) {
        ctx.setState(ClosedState(ctx));
      }
    } else {
      ctx.setState(ClosedState(ctx));
    }
  }
}
